src/simulation.py

```diff
@@ -0,0 +1,233 @@
+from src.custom_types import Solution, Vertex, RGBA, Canvas
+from src.reconstruction import polygon_init, polygon_mutate
+from src.visualize import add_polygon, visualize_canvas
+from src.loss import sad
+from numpy import array
+from copy import copy, deepcopy
+import numpy as np
+
+
+class Simulation:
+    def __init__(self, **kwargs):
+        """
+        Init simulation.
+        Keywords:
+            - Base image
+            - Output image
+            - Max generations
+            - Max iterations
+            - Stagnation limit
+            - Number evaluations
+            - Height
+            - Width
+            - Number Verticies
+        """
+
+        self.base_image = kwargs.get("b_image")
+        self.output_image = kwargs.get("o_image")
+        self.max_generations: int = kwargs.get("m_gen", 10)
+        self.max_iterations: int = kwargs.get("m_iter", 10)
+        self.stagnation_limit: int = kwargs.get("stag_lim", 10)
+        self.n_verticies: int = kwargs.get("n_vert", 3)
+        self.num_evals: int = kwargs.get("n_evals", 50000)
+        # NOTE: this value is from the paper
+
+        # Derived class variables
+        self.height, self.width = self.base_image.shape
+        self.canvas = Canvas(list())
+        self.counter = 0
+
+        self.canvas = self.create_polygon(self.canvas)
+
+    def update_probabilities(
+        self,
+    ):
+        """
+        Choose a polygon to mutate weighted by the sequence probabilities
+        """
+        num = self.canvas.how_many()
+
+        # Generate Geometric series of probabilities for the sequence
+        probabilities = array([1 / (2 ** (1 + x)) for x in range(num)])
+
+        # Normalize the probabilities to 1
+        n_probabilities = probabilities / np.sum(probabilities)
+        self.probabilities = n_probabilities
+        return n_probabilities
+
+    def norm_opti_probs(
+        self,
+    ):
+        """
+        Create a uniform distribution with mg elements, where mg is the max
+        generations / max number of polygons
+        """
+        self.probabilities = [1 / self.max_generations] * self.max_generations
+        return self.probabilities
+
+    def eval_loss(self, image):
+        """
+        Evaluate an image to the base_image and return the SAD
+        """
+        return sad(self.base_image, image)
+
+    def cc_loss(self, parent, child):
+        """
+        Compute and compare loss
+        """
+
+        # compute the loss of the child iteration with the parent
+        l_parent = self.eval_loss(parent)
```

Comment:
We could keep an attribute for keeping track of `l_parent` so that we don't compute it for every singe `l_child`.